### PR TITLE
Set Gradle Proxy Script: Fix Home Variable and Add no_proxy

### PR DESCRIPTION
**Pull request by jens-erdmann**
In https://github.com/oss-review-toolkit/ort/commit/c6b2818af7ed7b960ed8977449314eb4fc6ed783 the docker file was fixed to properly export the gradle
home variable for the gradle application to use.
The set_gradle_proxy.sh script needs to be modified to follow this
change. Up until now it was expecting gradle properties to reside in
$HOME while gradle is now expecting it at GRADLE_USER_HOME.

Whithout this adaptation gradle will fail with a
java.net.UnknownHostException error.
Complete error:
```
#17 1.561 Setting HTTP proxy http://proxy.example.com:3128 for Gradle in file '/root/.gradle/gradle.properties'...
#17 1.564 Setting HTTPS proxy http://proxy.example.com:3128 for Gradle in file '/root/.gradle/gradle.properties'...
#17 1.693 Downloading https://services.gradle.org/distributions/gradle-7.4.1-bin.zip
#17 1.910
#17 1.920 Exception in thread "main"    : services.gradle.org
#17 1.920       at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:229)
#17 1.920       at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
#17 1.920       at java.base/java.net.Socket.connect(Socket.java:609)
#17 1.920       at java.base/sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:300)
#17 1.921       at java.base/sun.net.NetworkClient.doConnect(NetworkClient.java:177)
#17 1.921       at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:474)
#17 1.921       at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:569)
#17 1.921       at java.base/sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:266)
#17 1.921       at java.base/sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:373)
#17 1.921       at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:203)
#17 1.921       at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1187)
#17 1.921       at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1081)
#17 1.922       at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:189)
#17 1.922       at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1592)
#17 1.922       at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1520)
#17 1.922       at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:250)
#17 1.922       at org.gradle.wrapper.Download.downloadInternal(Download.java:100)
#17 1.922       at org.gradle.wrapper.Download.download(Download.java:80)
#17 1.922       at org.gradle.wrapper.Install$1.call(Install.java:82)
#17 1.923       at org.gradle.wrapper.Install$1.call(Install.java:62)
#17 1.923       at org.gradle.wrapper.ExclusiveFileAccessManager.access(ExclusiveFileAccessManager.java:69)
#17 1.923       at org.gradle.wrapper.Install.createDist(Install.java:62)
#17 1.923       at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:107)
#17 1.923       at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:63)
```

Also adding functionality to process no_proxy environment variable and to set gradle properties accordingly.

### CODE REVIEW

File: scripts/set_gradle_proxy.sh [hunk 1]
```diff
@@ -53,7 +53,7 @@ writeProxyStringToGradleProps () {
 }
 
 writeProxyEnvToGradleProps () {
-    local GRADLE_PROPS="$HOME/.gradle/gradle.properties"
+    local GRADLE_PROPS="$GRADLE_USER_HOME/gradle.properties"
```
Comment (sschuberth) [inline scripts/set_gradle_proxy.sh:69]: Commit message nits:
- Just use the literal filename (`set_gradle_proxy.sh`) as the prefix in the title.
- In exchange, drop "set_gradle_proxy.sh" from the sentence in the commit message body.- s/docker/Docker/
- s/gradle/Gradle/
- s/GRADLE_USER_HOME/$GRADLE_USER_HOME/
- Consider putting "quoted code" (environment variables, exception names) into backticks like in Markdown.

File: scripts/set_gradle_proxy.sh [hunk 2]
```diff
@@ -52,6 +52,19 @@ writeProxyStringToGradleProps () {
 	EOF
 }
 
+writeNoProxyEnvToGradleProps() {
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply scripts/set_gradle_proxy.sh:55]: Please use a single space between function name and `()`.

File: scripts/set_gradle_proxy.sh [hunk 3]
```diff
@@ -52,6 +52,19 @@ writeProxyStringToGradleProps () {
...
+    local HOSTS=$1
+    local FILE=$2
+
+    grep -qF "systemProp.http.nonProxyHosts" $FILE 2>/dev/null && return 1
+
+    # Gradle / JVM expects a list separated by pipes instead of the comma that
+    # is used in shell environments
+    PREPED_HOSTS="$(echo ${HOSTS} | sed -e 's/,/|/g')"
+
+    echo "systemProp.http.nonProxyHosts=${PREPED_HOSTS}" >> $FILE
```
Comment (sschuberth) [inline scripts/set_gradle_proxy.sh:65]: We're already making use of Bash's string manipulation capabilities, so we could simply use `${HOSTS//,/|}` here.

But before that, could you do me a favor and add a separate commit before this one that changes the shebang lines in `import_proxy_certs.sh` and `set_gradle_proxy.sh` from `#!/bin/sh` to `#!/bin/bash`?

**Review by sschuberth — CHANGES_REQUESTED** [review]
Thanks for this! Just some minor things to address.

File: scripts/set_gradle_proxy.sh (hunk 3 above)
Comment (jens-erdmann) [reply scripts/set_gradle_proxy.sh:65]: > We're already making use of Bash's string manipulation capabilities, so we could simply use ${HOSTS//,/|} here.

Last time I tried I got a "bad substitution" error. But maybe that was related to the shebang.

Comment (jens-erdmann) [reply scripts/set_gradle_proxy.sh:65]: Done and done.

File: scripts/set_gradle_proxy.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (jens-erdmann) [reply scripts/set_gradle_proxy.sh]: Done and done.

File: scripts/set_gradle_proxy.sh (hunk 1 above)
Comment (jens-erdmann) [reply scripts/set_gradle_proxy.sh:69]: Done

Comment (sschuberth) [reply scripts/set_gradle_proxy.sh:69]: It just occurred to me that this script should also work outside of the `Dockerfile` context, where the `GRADLE_USER_HOME` variable might not be set. So we should probably use `${GRADLE_USER_HOME:-$HOME/.gradle}` here.

File: scripts/import_proxy_certs.sh [hunk 4]
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
```
Comment (sschuberth) [inline scripts/import_proxy_certs.sh:1]: Nits:
- Maybe make this the first commit in the PR, as it fixes something that was already wrong before.
- If referring to more than one file, consider using the parent directory as prefix in the title instead of list all files: `scripts: ...`.

**Review by sschuberth — CHANGES_REQUESTED** [review]
One more round, sorry 😁

File: scripts/set_gradle_proxy.sh (hunk 3 above)
Comment (jens-erdmann) [reply scripts/set_gradle_proxy.sh]: Done

**Review by sschuberth — DISMISSED** [review]

File: scripts/import_proxy_certs.sh (hunk 4 above)
Comment (sschuberth) [reply scripts/import_proxy_certs.sh:1]: I was referring to simply reordering the commits in this PR, making the Shebang fix the first in the history of this branch, but that was completely optional anyway.

**Review by sschuberth — DISMISSED** [review]
